src/crimson/admin/admin_socket.cc

```diff
@@ -0,0 +1,646 @@
+  if (parsed.m_cmd_seq_len == parsed.m_cmd.length())
+    return true;
+
+  logger().info("{}: validating {} against:{}", __func__, command_text,
+                parsed.m_api->cmddesc);
+
+  stringstream os;  // for possible validation error messages
+  try {
+    // validate_cmd throws on some syntax errors
+    if (validate_cmd(nullptr, parsed.m_api->cmddesc, parsed.m_parameters, os)) {
+      return true;
+    }
+  } catch (std::exception& e) {
+    logger().error("{}: validation failure ({} : {}) {}", __func__,
+                   command_text, parsed.m_cmd, e.what());
+  }
+
+  os << "error: command validation failure ";
+  logger().error("{}: validation failure (incoming:{}) {}", __func__,
+                 command_text, os.str());
+  out.append(os);
+  return false;
+}
+
+seastar::future<> AdminSocket::finalize_response(
+  seastar::output_stream<char>& out, ceph::bufferlist&& msgs)
+{
+  string outbuf_cont = msgs.to_str();
+  if (outbuf_cont.empty()) {
+    outbuf_cont = " {} ";
+  }
+  uint32_t response_length = htonl(outbuf_cont.length());
+  logger().info("asok response length: {}", outbuf_cont.length());
+
+  return out.write((char*)&response_length, sizeof(uint32_t))
+    .then([&out, outbuf_cont] { return out.write(outbuf_cont.c_str()); });
+}
+
+seastar::future<> AdminSocket::execute_line(std::string cmdline,
+                                            seastar::output_stream<char>& out)
+{
+  return seastar::with_shared(
+    servers_tbl_rwlock, [this, cmdline, &out]() mutable {
+      ceph::bufferlist msgs;
+      maybe_parsed_t oparsed;
+
+      bool ok_to_execute{ false };
+      try {
+        oparsed = parse_cmd(cmdline, msgs);
+        if (oparsed.has_value())
+          ok_to_execute = validate_command(*oparsed, cmdline, msgs);
+      } catch (...) {
+      };
+
+      if (!ok_to_execute) {
+        return finalize_response(out, std::move(msgs));
+      }
+
+      return seastar::do_with(
+        std::move(*oparsed), std::move(msgs),
+        [this, &out](auto&& parsed, auto&& out_buf) {
+          return (parsed.m_hook->call(parsed.m_cmd, parsed.m_parameters,
+                                      parsed.m_format, out_buf))
+            .then_wrapped([](auto&& fut) -> seastar::future<bool> {
+              if (fut.failed()) {
+                // add 'failed' to the contents of out_buf? not what
+                // happens in the old code
+                return seastar::make_ready_future<bool>(false);
+              } else {
+                return std::move(fut);
+              }
+            })
+            .then([this, &out, &out_buf](auto call_res) {
+              return finalize_response(out, std::move(out_buf));
+            });
+        });
+    });
+}
+
+seastar::future<> AdminSocket::handle_client(seastar::input_stream<char>& inp,
+                                             seastar::output_stream<char>& out)
+{
+  // \todo handle old protocol (see original code)
```

Comment:
please note doxygen does not render comments like this.

Reply:
will fix to /// \todo (which is rendered, at least with standard configuration of doxygen)